enclave_build/src/image_manager.rs

```diff
@@ -0,0 +1,312 @@
+
+use tempfile::NamedTempFile;
+use tokio::runtime::Runtime;
+
+#[derive(Debug)]
+pub enum Error {
+    ImagePull(crate::pull::Error),
+    Cache(crate::cache::Error),
+    Extract(String),
+    ImageConvert(crate::image::Error),
+    ImageConfigMissing,
+    EntryPoint,
+    Tempfile,
+    Runtime,
+}
+
+impl std::fmt::Display for Error {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match self {
+            Error::ImagePull(err) => write!(
+                f,
+                "Could not pull image data from the remote registry: {:?}",
+                err
+            ),
+            Error::Cache(err) => write!(f, "Cache error: {:?}", err),
+            Error::Extract(msg) => write!(f, "Failed to extract expressions from image: {}", msg),
+            Error::ImageConvert(err) => write!(f, "Image convert error: {:?}", err),
+            Error::ImageConfigMissing => write!(f, "Image configuration could not be extracted."),
+            Error::EntryPoint => write!(f, "Entrypoint missing."),
+            Error::Tempfile => write!(f, "Error when handling a tempfile"),
+            Error::Runtime => write!(f, "Runtime error."),
+        }
+    }
+}
+
+impl From<crate::pull::Error> for Error {
+    fn from(err: crate::pull::Error) -> Self {
+        Error::ImagePull(err)
+    }
+}
+
+impl From<crate::cache::Error> for Error {
+    fn from(err: crate::cache::Error) -> Self {
+        Error::Cache(err)
+    }
+}
+
+impl std::error::Error for Error {}
+
+/// Trait which provides an interface for handling images.
+pub trait ImageManager {
+    fn get_image_name(&self) -> &String;
+    /// Pulls the image from remote and stores it in the local cache.
+    fn pull_image(&mut self) -> result::Result<(), Box<dyn std::error::Error>>;
+    /// Builds an image locally (from a Dockerfile in case of Docker).
+    fn build_image(&self, dockerfile_dir: String)
+        -> result::Result<(), Box<dyn std::error::Error>>;
+    /// Inspects the image and returns its metadata in the form of a JSON Value.
+    fn inspect_image(&mut self) -> result::Result<serde_json::Value, Box<dyn std::error::Error>>;
+    /// Returns the architecture of the image.
+    fn architecture(&mut self) -> result::Result<String, Box<dyn std::error::Error>>;
+    /// Returns two temp files containg the CMD and ENV expressions extracted from the image,
+    /// in this order.
+    fn load(
+        &mut self,
+    ) -> result::Result<(NamedTempFile, NamedTempFile), Box<dyn std::error::Error>>;
+}
+
+pub type Result<T> = std::result::Result<T, Error>;
+
+pub struct OciImageManager {
+    /// Name of the container image.
+    image_name: String,
+    /// Have the cache as an option in order to not stop the CLI if there is a cache creation
+    /// error (the image will simply be pulled in this case, instead of being fetched from cache)
+    cache: Option<CacheManager>,
+}
+
+impl ImageManager for OciImageManager {
+    fn get_image_name(&self) -> &String {
```

Comment:
Getters in Rust usually don't include the `get_` part in their name: https://rust-lang.github.io/api-guidelines/naming.html#getter-names-follow-rust-convention-c-getter.